DSpace/dspace-angular
# Features support

## References
This PR connects to #664

## Description
This PR adds the ability to check authorizations for features through REST calls.
It also adds an implementation using the `loginOnBehalfOf` and `administratorOf` feature for disabling the impersonate button and hiding / restricting access to admin pages respectively.

## Instructions for Reviewers
Changes made:
- Added models and services for `Feature` and `Authorization` 
-- The `AuthorizationDataService` contains a method `isAuthorized()`, which returns an observable emitting true if the user is authorized for the provided feature and object, false if he/she isn't authorized. (A user is considered authorized if the list of authorizations returned by the REST API contains at least one Authorization)
-- To pass the ID of a feature to check authorization for, we're making use of the new enum `FeatureID`.
-- If no object is provided with `isAuthorized()`, it will use the repository's `Site`
-- If no user UUID is provided with `isAuthorized()`, it will use the currently authenticated `EPerson` (or undefined for anonymous)
- Added implementation for `canImpersonate$` on `EPersonFormComponent`. This will disable the impersonate button if the current user is not authorized to do so.
- Added `SiteAdministratorGuard` extending from abstract `FeatureAuthorizationGuard`. The `SiteAdministratorGuard` is used to restrict access to certain modules for users that aren't Site administrators (configured this guard for route `/admin`)
-- These guards will redirect the user to a new `UnauthorizedComponent`. This is similar to the `PageNotFoundComponent`, but displays messages for unauthorized (401)
- Modified `AdminSidebarComponent` to hide administrator routes from the sidebar, making use of the new authorization features.

How to test:
- Login as a non-administrator and open the sidebar. All administrator links (e.g. metadata registry) should be hidden.
- Login as administrator and open the sidebar. All sidebar options should be visible.
- Login as administrator and visit `/admin/access-control/epeople`. Edit an EPerson. The impersonate button should be enabled/disabled depending on the administrator's `loginOnBehalfOf` rights.
- Try to access an administrator feature (e.g.  `/admin/access-control/epeople`) as anonymous. You should be redirected to `/unauthorized`.
- Try to access an administrator feature as a non-administrator. You should be redirected to `/unauthorized`. **Warning**: If this redirects you to the homepage instead, this is due to #711 .

Note: We've tried to maintain the original URLs the user is being redirected from for unauthorized pages, but Guards in angular don't allow this.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_
    * Include tests for different user types (if behavior differs), including: (1) Anonymous user, (2) Logged in user (non-admin), and (3) Administrator.
    * Include tests for error scenarios, e.g. when errors/warnings should appear (or buttons should be disabled).
    * For bug fixes, include a test that reproduces the bug and proves it is fixed. For clarity, it may be useful to provide the test in a separate commit from the bug fix.
